File: app/src/main/java/be/digitalia/fosdem/db/entities/EventEntity.java
package be.digitalia.fosdem.db.entities;

import java.util.Date;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;
import be.digitalia.fosdem.db.converters.NullableDateTypeConverters;
import be.digitalia.fosdem.model.Event;

@Entity(tableName = EventEntity.TABLE_NAME, indices = {
		@Index(value = {"day_index"}, name = "event_day_index_idx"),
		@Index(value = {"start_time"}, name = "event_start_time_idx"),
		@Index(value = {"end_time"}, name = "event_end_time_idx"),
		@Index(value = {"track_id"}, name = "event_track_id_idx")
})
public class EventEntity {

	public static final String TABLE_NAME = "events";

	@PrimaryKey
	private final long id;
	@ColumnInfo(name = "day_index")
	private final int dayIndex;
	@ColumnInfo(name = "start_time")
	@TypeConverters({NullableDateTypeConverters.class})
	private final Date startTime;
	@ColumnInfo(name = "end_time")
	@TypeConverters({NullableDateTypeConverters.class})
	private final Date endTime;
	@ColumnInfo(name = "room_name")
	private final String roomName;
	private final String slug;
	@ColumnInfo(name = "track_id")
	private final long trackId;
	@ColumnInfo(name = "abstract")
	private final String abstractText;
	private final String description;

	public EventEntity(Event event) {
		this(event.getId(), event.getDay().getIndex(), event.getStartTime(), event.getEndTime(), event.getRoomName(),
				event.getSlug(), event.getTrack().getId(), event.getAbstractText(), event.getDescription());
	}

	public EventEntity(long id, int dayIndex, Date startTime, Date endTime, String roomName,
					   String slug, long trackId, String abstractText, String description) {
		this.id = id;
		this.dayIndex = dayIndex;
		this.startTime = startTime;
		this.endTime = endTime;
		this.roomName = roomName;
		this.slug = slug;
		this.trackId = trackId;
		this.abstractText = abstractText;
		this.description = description;
	}

	public long getId() {
		return id;
	}

	public int getDayIndex() {
		return dayIndex;
	}

	public Date getStartTime() {
		return startTime;
	}

	public Date getEndTime() {
		return endTime;
	}

	public String getRoomName() {
		return roomName;
	}

	public String getSlug() {
		return slug;
	}

	public long getTrackId() {
		return trackId;
	}

	public String getAbstractText() {
		return abstractText;
	}

	public String getDescription() {
		return description;
	}
}

File: app/src/main/java/be/digitalia/fosdem/db/entities/EventEntity.kt
package be.digitalia.fosdem.db.entities

import androidx.room3.ColumnInfo
import androidx.room3.ColumnTypeConverters
import androidx.room3.Entity
import androidx.room3.Index
import androidx.room3.PrimaryKey
import be.digitalia.fosdem.db.converters.NullableInstantTypeConverters
import be.digitalia.fosdem.db.converters.NullableZoneOffsetTypeConverters
import java.time.ZoneOffset
import kotlin.time.Instant

@Entity(
    tableName = EventEntity.TABLE_NAME,
    indices = [
        Index(value = ["day_index"], name = "event_day_index_idx"),
        Index(value = ["start_time"], name = "event_start_time_idx"),
        Index(value = ["end_time"], name = "event_end_time_idx"),
        Index(value = ["track_id"], name = "event_track_id_idx")
    ]
)
class EventEntity(
        @PrimaryKey
        val id: Long,
        @ColumnInfo(name = "day_index")
        val dayIndex: Int,
        @ColumnInfo(name = "start_time")
        @field:ColumnTypeConverters(NullableInstantTypeConverters::class)
        val startTime: Instant?,
        @ColumnInfo(name = "start_time_offset")
        @field:ColumnTypeConverters(NullableZoneOffsetTypeConverters::class)
        val startTimeOffset: ZoneOffset?,
        @ColumnInfo(name = "end_time")
        @field:ColumnTypeConverters(NullableInstantTypeConverters::class)
        val endTime: Instant?,
        @ColumnInfo(name = "room_name")
        val roomName: String?,
        val url: String?,
        @ColumnInfo(name = "track_id")
        val trackId: Long,
        @ColumnInfo(name = "abstract")
        val abstractText: String?,
        val description: String?,
        @ColumnInfo(name = "feedback_url")
        val feedbackUrl: String?
) {
    companion object {
        const val TABLE_NAME = "events"
    }
}